File: R/navigation.R
##
## Copyright (c) 2018-2026 BigOmics Analytics SA. All rights reserved.
##

## Server-side navigation. Each of these sends a custom message handled in
## srcjs/navigation.js, or calls one of the globals defined there.
##
## The `bigdash.` prefix predates these living in the package; it is kept so
## that calling code reads the same whether it dot-sources them or imports
## the package.

#' @keywords internal
send_nav <- function(session, type, value = NULL) {
  ## Same class list shiny:::validate_session_object accepts. Inside
  ## moduleServer the session is a bare `session_proxy`, which does NOT
  ## inherit from ShinySession, so checking for that alone rejects every
  ## call made from within a module.
  if (!inherits(session, c("ShinySession", "MockShinySession", "session_proxy"))) {
    stop("`session` must be a Shiny session object", call. = FALSE)
  }
  session$sendCustomMessage(type, if (is.null(value)) list() else list(value = value))
}

#' Resolve the [bigPage()] id `session` belongs to, for scoping nav messages.
#'
#' `session$ns(NULL)` is `""`/`character(0)` for the top-level session and
#' the module id for a `moduleServer()` session_proxy -- which is exactly the
#' `id` convention [bigPage()]/[sidebar()]/[settings()] already use (see
#' [scoped_id()]). Returning `NULL` for the top-level case (rather than `""`)
#' matters: `send_nav()` treats a `NULL` value as "no value field at all",
#' which the JS side's `scopedId()` then reads as the unscoped default,
#' matching pre-existing, page-wide behaviour exactly.
#'
#' @keywords internal
nav_scope_id <- function(session) {
  if (is.null(session)) {
    return(NULL)
  }
  id <- session$ns(NULL)
  if (length(id) == 0 || !nzchar(id)) {
    return(NULL)
  }
  id
}

#' Select a tab
#' @param session Shiny session.
#' @param selected `data-target` of the tab to select.
#' @export
bigdash.selectTab <- function(session, selected) {
  send_nav(session, "bigdash-select-tab", selected)
}

#' Finish wiring a [bigPage()] root inserted into the DOM after page load
#'
#' bigdash's one-time page-ready setup (settings-panel relocation, first-tab
#' selection) only ever sees the roots present in the DOM at that moment.
#' Call this once, right after inserting a new [bigPage()] at runtime (e.g.
#' via `bslib::nav_insert()` or `shiny::insertUI()` to lazily load a
#' module's UI on demand), to run that same per-root setup for it. Click
#' handlers (sidebar collapse, settings drawer, tab switching) do not need
#' this -- they are bound delegated and already cover new roots.
#'
#' Deferred via `session$onFlush()` so it is sent after (not possibly
#' before) the message that actually inserts the new root's markup, as long
#' as this is called after the insertion call in the same reactive tick --
#' `onFlush` callbacks run, and so their messages are sent to the client, in
#' registration order.
#'
#' @param session Shiny session.
#' @param id The `id` passed to the newly-inserted [bigPage()].
#' @export
bigdash.initRoot <- function(session, id) {
  session$onFlush(function() {
    send_nav(session, "bigdash-init-root", id)
  }, once = TRUE)
}

#' Reveal the sidebar menu after the landing page
#' @param session Shiny session.
#' @export
bigdash.showTabs <- function(session) {
  send_nav(session, "show-tabs", nav_scope_id(session))
}

## ------------------- sideBar ---------------------------
#' Open the sidebar
#'
#' @param session Shiny session that identifies which [bigPage()] instance to
#'   target. Defaults to the current reactive domain, so existing no-argument
#'   calls keep working unchanged -- at the top level that resolves to the
#'   default/unscoped `bigPage()`, and inside a `moduleServer()` it resolves
#'   to that module's own scoped `bigPage()`. Pass it explicitly (or leave
#'   the default) from inside a module to target that module's sidebar
#'   rather than the page-wide default. `NULL` (no reactive domain found, or
#'   passed explicitly) falls back to the pre-existing, unscoped behaviour.
#' @export
bigdash.openSidebar <- function(session = shiny::getDefaultReactiveDomain()) {
  if (is.null(session)) {
    shinyjs::runjs("sidebarOpen()")
    return(invisible(NULL))
  }
  send_nav(session, "bigdash-open-sidebar", nav_scope_id(session))
}

#' Close the sidebar
#' @inheritParams bigdash.openSidebar
#' @export
bigdash.closeSidebar <- function(session = shiny::getDefaultReactiveDomain()) {
  if (is.null(session)) {
    shinyjs::runjs("sidebarClose()")
    return(invisible(NULL))
  }
  send_nav(session, "bigdash-close-sidebar", nav_scope_id(session))
}

#' Collapse the sidebar back to its initial state
#' @inheritParams bigdash.openSidebar
#' @export
bigdash.unloadSidebar <- function(session = shiny::getDefaultReactiveDomain()) {
  if (is.null(session)) {
    shinyjs::runjs("unloadSidebar()")
    return(invisible(NULL))
  }
  send_nav(session, "bigdash-unload-sidebar", nav_scope_id(session))
}

#' Open or close the sidebar
#' @param state `TRUE` to open, `FALSE` to close.
#' @inheritParams bigdash.openSidebar
#' @export
bigdash.toggleSidebar <- function(state, session = shiny::getDefaultReactiveDomain()) {
  if (state) bigdash.openSidebar(session)
  if (!state) bigdash.closeSidebar(session)
}

## ------------------- settingsBar ---------------------------
#' Open the settings drawer
#' @param lock Ignored, kept so existing calls keep working. The drawer no
#'   longer closes itself when the pointer leaves, so it is always "locked".
#' @inheritParams bigdash.openSidebar
#' @export
bigdash.openSettings <- function(lock = TRUE, session = shiny::getDefaultReactiveDomain()) {
  if (is.null(session)) {
    shinyjs::runjs("settingsOpen()")
    return(invisible(NULL))
  }
  send_nav(session, "bigdash-open-settings", nav_scope_id(session))
}

#' Close the settings drawer
#' @inheritParams bigdash.openSidebar
#' @export
bigdash.closeSettings <- function(session = shiny::getDefaultReactiveDomain()) {
  if (is.null(session)) {
    shinyjs::runjs("settingsClose()")
    return(invisible(NULL))
  }
  send_nav(session, "bigdash-close-settings", nav_scope_id(session))
}

## --------------------menuItem --------------------------
#' Show a sidebar menu item
#' @param session Shiny session.
#' @param item `data-target` of the menu item.
#' @export
bigdash.showMenuItem <- function(session, item) {
  send_nav(session, "bigdash-show-menuitem", item)
}

#' Hide a sidebar menu item
#' @param session Shiny session.
#' @param item `data-target` of the menu item.
#' @export
bigdash.hideMenuItem <- function(session, item) {
  send_nav(session, "bigdash-hide-menuitem", item)
}

#' Show or hide a sidebar menu item
#' @param session Shiny session.
#' @param item `data-target` of the menu item.
#' @param state `TRUE` to show, `FALSE` to hide.
#' @export
bigdash.toggleMenuItem <- function(session, item, state) {
  if (state) bigdash.showMenuItem(session, item)
  if (!state) bigdash.hideMenuItem(session, item)
}

## --------------------BigTab --------------------------
#' Show a tab and its menu item
#' @param session Shiny session.
#' @param tab `data-name` of the tab.
#' @export
bigdash.showTab <- function(session, tab) {
  send_nav(session, "bigdash-show-tab", tab)
  bigdash.showMenuItem(session, tab)
}

#' Hide a tab and its menu item
#' @param session Shiny session.
#' @param tab `data-name` of the tab.
#' @export
bigdash.hideTab <- function(session, tab) {
  send_nav(session, "bigdash-hide-tab", tab)
  bigdash.hideMenuItem(session, tab)
}

#' Show or hide a tab
#' @param session Shiny session.
#' @param tab `data-name` of the tab.
#' @param state `TRUE` to show, `FALSE` to hide.
#' @export
bigdash.toggleTab <- function(session, tab, state) {
  if (state) bigdash.showTab(session, tab)
  if (!state) bigdash.hideTab(session, tab)
}

#' Remove a tab, its settings pane and its menu item
#' @param session Shiny session.
#' @param tab `data-name` of the tab.
#' @export
bigdash.removeTab <- function(session, tab) {
  send_nav(session, "bigdash-remove-tab", tab)
}

#' Hide a sidebar menu group by its id
#'
#' Hides the group's header/chevron/divider and closes its collapse body.
#' The body reappears closed (not force-expanded) the next time
#' [bigdash.showMenuElement()] reveals the group -- same as a freshly
#' rendered [sidebarMenu()], the user re-expands it by clicking.
#'
#' @param session Shiny session.
#' @param id The target [sidebarMenu()]'s own `id`. [sidebarMenu()]
#'   auto-generates one if you don't pass `id` yourself, so targeting a
#'   specific group from the server requires giving it an explicit,
#'   stable `id` in the UI first: `sidebarMenu(..., id = "reports-menu")`.
#'
#'   If nothing matches `id`, this falls back to matching the group's
#'   visible label instead (the pre-`id` behaviour) -- purely so callers
#'   written before [sidebarMenu()] took an `id` keep working unchanged.
#'   New code should always give its [sidebarMenu()] an explicit `id` and
#'   target that; matching by label breaks the moment two groups share a
#'   label or one gets renamed/translated.
#'
#' @export
bigdash.hideMenuElement <- function(session, id) {
  send_nav(session, "bigdash-hide-menu-element", id)
}

#' Show a sidebar menu group by its id
#' @inheritParams bigdash.hideMenuElement
#' @export
bigdash.showMenuElement <- function(session, id) {
  send_nav(session, "bigdash-show-menu-element", id)
}

#' Filter visible tabs and menu items
#'
#' Shows only the specified tabs (and their corresponding sidebar menu items).
#' Hides all other menu items and their associated tab canvases (`big-tab`
#' elements). Useful for dynamic filtering of available boards/views.
#'
#' The currently active tab stays selected if it is still among `tabs`;
#' otherwise the first tab in `tabs` is selected. This avoids bouncing the
#' user back to `tabs[[1]]` on every call when their current tab was never
#' actually filtered out.
#'
#' @param session Shiny session.
#' @param tabs Character vector of tab names (`data-name` / `target` values
#'   from [bigTabItem()], [sidebarItem()], etc.).
#' @param id Namespace id of the [bigPage()] instance to filter, as given to
#'   [bigPage()]/[bigTabs()]/[sidebar()] in the UI. Only needed when several
#'   `bigPage()`s render in parallel from the same server, each with its own
#'   `id` (not nested) -- omit it for the default, single-`bigPage()` case.
#' @export
bigdash.filterTabs <- function(session, tabs, id = BIGDASH_DEFAULT_ID) {
  if (!length(tabs)) {
    return(invisible(NULL))
  }
  if (!inherits(tabs, "character")) {
    tabs <- as.character(tabs)
  }
  tabs <- unique(tabs)

  ## The nav input is set at the root, under the bigTabs id ("qsee-nav").
  ## Read from inside module "qsee" that is input$nav; read from the app
  ## server it is input[["qsee-nav"]]. Strip the caller's own prefix when
  ## it is there. (Same lookup as bd_active_tab(), kept private here so
  ## this doesn't reach across into an unrelated module.)
  nav <- scoped_id(id, "nav")
  prefix <- session$ns("")
  if (nzchar(prefix) && startsWith(nav, prefix)) {
    nav <- substring(nav, nchar(prefix) + 1L)
  }
  current <- shiny::isolate(session$input[[nav]])
  select <- if (!is.null(current) && current %in% tabs) current else tabs[[1]]

  send_nav(session, "bigdash-filter-tabs", list(id = id, keep = tabs, select = select))
}

File: srcjs/navigation.js
import 'jquery';
import 'shiny';
import { refreshHelp, initSidebarRoot } from './sidebar';
import { settingsExpand, settingsCollapse, initSettingsRoot, moveSettingsForRoot } from './settings';
import { DEFAULT_ID, scopedId } from './scope';

// Programmatic counterpart to sidebar.js / settings.js: these drive the same
// DOM those modules wire up, so the server can open, close and hide things.
// They hang off `window` because shinyjs::runjs evaluates in global scope.

// Every sidebarMenu() group should disappear entirely -- header, chevron,
// hr, collapse body -- the moment hiding/showing leaves it with zero visible
// sidebarMenuItem()s; a header that expands to an empty box is never useful,
// so this applies unconditionally, not just to promote_single = TRUE groups.
// A sidebarMenu(promote_single = TRUE) group additionally reads as a flat,
// top-level item -- no header/chevron, no indent -- whenever exactly one
// item is left visible. Both are re-run after every message that can change
// item-level visibility (below), so they stay correct through any number of
// filterTabs()/hideMenuItem() calls across the session, not just at initial
// render.
//
// `.css('display')` (not jQuery's `:visible`) on purpose: `:visible` also
// factors in the `.collapse` ancestor's own open/closed state, which would
// make every item in a currently-collapsed group read as "not visible" even
// though it's just accordion-closed, not filtered out.
const refreshMenuPromotion = () => {
	$('.sidebar-menu').each((i, headerLink) => {
		const $headerLink = $(headerLink);
		const $header = $headerLink.closest('p');
		const $hr = $header.next('hr');
		const $collapse = $(`#${$headerLink.data('target')}`);
		if ($collapse.length === 0) return;

		const visibleCount = $collapse
			.find('.tab-trigger')
			.filter((i, el) => $(el).css('display') !== 'none')
			.length;
		const empty = visibleCount === 0;
		const promote = !empty && visibleCount === 1 && $headerLink.data('promote-single') === true;

		$header.toggleClass('menu-empty', empty);
		$hr.toggleClass('menu-empty', empty);
		$collapse.toggleClass('menu-empty', empty);

		$header.toggleClass('menu-promoted', promote);
		$hr.toggleClass('menu-promoted', promote);
		$collapse.toggleClass('menu-promoted', promote);
	});
};

// Scoped so a bigPage() nested inside another one (e.g. a module embedded as
// a tab of the main app) can be driven on its own, without also touching
// every other bigPage() instance on the page. `id` defaults to DEFAULT_ID so
// the legacy, page-wide `window.*` helpers below keep their original,
// unscoped behaviour.
const unloadSidebar = (id = DEFAULT_ID) => {
	$(`#${scopedId(id, 'sidebar-container')} .sidebar-content`)
		.children()
		.each((index, el) => {
			if($(el).hasClass('collapse')){
				$(el).removeClass('show');
				$(el).hide();
				return;
			}
			if($(el).hasClass('nodisp')){
				$(el).hide();
				return;
			}
			$(el).show();
		});
        $(`#${scopedId(id, 'sidebar-help-container')}`).hide();
}

// the label click runs through setSidebarState, which calls refreshHelp; the
// help box is no longer shown or hidden from here, or it would override the
// "sidebar open AND tab has help" rule sidebar.js applies
const sidebarClose = (id = DEFAULT_ID) => {
    if($(`#${scopedId(id, 'sidebar-container')}`).hasClass('sidebar-expanded')) {
	$(`#${scopedId(id, 'sidebar-container')} .sidebar-label`).trigger('click');
    }
}

const sidebarOpen = (id = DEFAULT_ID) => {
    if($(`#${scopedId(id, 'sidebar-container')}`).hasClass('sidebar-collapsed')) {
	$(`#${scopedId(id, 'sidebar-container')} .sidebar-label`).trigger('click');
    }
}

// The panel is click-toggled from its label (settings.js), so drive that,
// the same way sidebarOpen/sidebarClose drive '.sidebar-label'. Testing for
// the absence of 'settings-collapsed' rather than the presence of
// 'settings-expanded' matters: a freshly rendered panel is open but carries
// neither class.
const settingsClose = (id = DEFAULT_ID) => {
	if(!$(`#${scopedId(id, 'settings-container')}`).hasClass('settings-collapsed'))
		$(`#${scopedId(id, 'settings-container')} .settings-label`).trigger('click');
}

const settingsOpen = (id = DEFAULT_ID) => {
	if($(`#${scopedId(id, 'settings-container')}`).hasClass('settings-collapsed'))
		$(`#${scopedId(id, 'settings-container')} .settings-label`).trigger('click');
}

export const handleNavigation = () => {
  // Unscoped globals: `shinyjs::runjs("sidebarOpen()")` etc (no bigPage() id
  // available) evaluate these in global scope, so they keep defaulting to
  // the page's default/top-level bigPage() instance, same as before.
  window.unloadSidebar = () => unloadSidebar();
  window.sidebarClose = () => sidebarClose();
  window.sidebarOpen = () => sidebarOpen();
  window.settingsClose = () => settingsClose();
  window.settingsOpen = () => settingsOpen();

  // Scoped counterparts: driven by `bigdash.open/closeSidebar(session)` and
  // `bigdash.open/closeSettings(session)`, which send the calling module's
  // id (or no value at all for the default/top-level session) as msg.value.
  Shiny.addCustomMessageHandler('bigdash-open-sidebar', (msg) => {
    sidebarOpen(msg.value);
  });

  Shiny.addCustomMessageHandler('bigdash-close-sidebar', (msg) => {
    sidebarClose(msg.value);
  });

  Shiny.addCustomMessageHandler('bigdash-unload-sidebar', (msg) => {
    unloadSidebar(msg.value);
  });

  Shiny.addCustomMessageHandler('bigdash-open-settings', (msg) => {
    settingsExpand(msg.value || DEFAULT_ID);
  });

  Shiny.addCustomMessageHandler('bigdash-close-settings', (msg) => {
    settingsCollapse(msg.value || DEFAULT_ID);
  });

  Shiny.addCustomMessageHandler('show-tabs', (msg) => {
	let id = msg.value || DEFAULT_ID;
	setTimeout(() => {
	$(`#${scopedId(id, 'sidebar-container')} .sidebar-content`)
		.children()
		.each((index, el) => {
      if ($(el).hasClass('collapse')) {
				$(el).removeClass('show');
				$(el).css({'display' : ''});
				return;
			}
			if($(el).hasClass('w-100')) {
				$(el).children().children()[[1]].classList.remove('fa-angle-down');
				$(el).children().children()[[1]].classList.add('fa-angle-right');
			}
      if (!$(el).hasClass('nodisp')) {
        $(el).show();
      }
		});

	refreshHelp(id);
	}, 1000);
  });

  Shiny.addCustomMessageHandler('bigdash-select-tab', (msg) => {
      $(`.tab-trigger[data-target=${msg.value}]`).trigger('click');
  });

  Shiny.addCustomMessageHandler('bigdash-hide-menuitem', (msg) => {
      $(`.tab-trigger[data-target=${msg.value}]`).hide();
      $(`.tab-trigger-hr[data-target=${msg.value}]`).hide();
      refreshMenuPromotion();
  });

  Shiny.addCustomMessageHandler('bigdash-show-menuitem', (msg) => {
      $(`.tab-trigger[data-target=${msg.value}]`).show();
	  $(`.tab-trigger-hr[data-target=${msg.value}]`).show();
      refreshMenuPromotion();
  });

  Shiny.addCustomMessageHandler('bigdash-hide-tab', (msg) => {
      $(`.big-tab[data-name=${msg.value}]`).hide();
  });

  Shiny.addCustomMessageHandler('bigdash-show-tab', (msg) => {
      $(`.big-tab[data-name=${msg.value}]`).show();
  });

  // Unlike bigdash-hide-tab/bigdash-show-tab (which only ever touch the one
  // tab named in msg.value), this walks every tab/menu-item currently in the
  // DOM so it can hide whatever is NOT in the kept list, not just show what
  // is -- the R side has no registry of "every tab that exists" to loop
  // hideTab() over, but the DOM does.
  //
  // Scoped to msg.value.id's own #big-tabs/#sidebar-container (scopedId()
  // resolves the default id to the same unscoped #big-tabs/#sidebar-container
  // selector as before, so this is a no-op change for the common
  // single-bigPage() case) -- needed so several bigPage()s rendered in
  // parallel with different ids from the same server don't filter each
  // other's tabs just because they happen to reuse a tab name.
  Shiny.addCustomMessageHandler('bigdash-filter-tabs', (msg) => {
      const id = msg.value.id || DEFAULT_ID;
      // A single-tab vector auto-unboxes to a bare string over the wire, not
      // a 1-element array -- normalise so .includes() below is always an
      // array membership check.
      const keep = [].concat(msg.value.keep || []);
      const select = msg.value.select;

      $(`#${scopedId(id, 'big-tabs')} .big-tab`).each((i, el) => {
          $(el).toggle(keep.includes($(el).data('name')));
      });
      $(`#${scopedId(id, 'sidebar-container')} .tab-trigger[data-target], #${scopedId(id, 'sidebar-container')} .tab-trigger-hr[data-target]`).each((i, el) => {
          $(el).toggle(keep.includes($(el).data('target')));
      });
      refreshMenuPromotion();

      // bigdash.filterTabs() already resolved this to the current tab (if
      // still visible) or tabs[[1]] -- re-clicking an already-active tab is
      // a harmless no-op (Shiny dedupes identical setInputValue calls), so
      // this is always safe to fire.
      if (select) {
          $(`#${scopedId(id, 'sidebar-container')} .tab-trigger[data-target=${select}]`).trigger('click');
      }
  });

  Shiny.addCustomMessageHandler('bigdash-remove-tab', (msg) => {
      $(`.big-tab[data-name=${msg.value}]`).remove();
	  $(`.tab-settings:has(a#${msg.value.slice(0, -4)}-options)`).remove();
      $(`[data-target=${msg.value}]`).hide();
      refreshMenuPromotion();
  });

  // Primarily targets the sidebarMenu() group whose own id is msg.value
  // (its .collapse body's DOM id, and what its header <a data-target=...>
  // points at) -- not just a `span:contains(text)` label match, which was
  // one false positive away from hiding the wrong group (any span
  // containing that substring anywhere on the page) and broke the moment a
  // label was translated or renamed.
  //
  // Falls back to that old label match when no id matches, purely for
  // compatibility with callers written before sidebarMenu() had a settable
  // `id` -- they have no choice but to pass the visible label, same as this
  // always required. New code should give its sidebarMenu() an explicit
  // `id` and target that instead.
  const findMenuHeader = (value) => {
      let $header = $(`.sidebar-menu[data-target="${value}"]`).closest('p');
      if ($header.length === 0) {
          $header = $(`.sidebar-menu span:contains(${value})`).closest('p');
      }
      return $header;
  };

  Shiny.addCustomMessageHandler('bigdash-hide-menu-element', (msg) => {
      const $header = findMenuHeader(msg.value);
      // .nodisp marks the header as deliberately hidden so
      // show-tabs/unloadSidebar's landing-page reveal (navigation.js/
      // sidebar.js) leaves it alone, same contract every other
      // manually-hidden element here already relies on.
      $header.hide().addClass('nodisp');
      $header.next('hr').hide();
      // Close the body too, not just the header -- otherwise a group that
      // was open when hidden leaves its items sitting on screen with no
      // header left to click closed. Resolved from the header's own
      // data-target, not msg.value directly -- those differ when msg.value
      // matched via the legacy label fallback above.
      const collapseId = $header.find('.sidebar-menu').data('target');
      if (collapseId) {
          $(`#${collapseId}`).removeClass('show').hide();
      }
  });

  Shiny.addCustomMessageHandler('bigdash-show-menu-element', (msg) => {
      const $header = findMenuHeader(msg.value);
      $header.show().removeClass('nodisp');
      $header.next('hr').show();
      // Clear the inline display hide-menu-element set above so the
      // .collapse's own bootstrap class (show/not) governs it again --
      // reappears closed, same as a freshly rendered sidebarMenu(), not
      // forced open.
      const collapseId = $header.find('.sidebar-menu').data('target');
      if (collapseId) {
          $(`#${collapseId}`).css('display', '').removeClass('show');
      }
  });

  // Finish wiring a bigPage() root that was inserted into the DOM after
  // page load (e.g. via bslib::nav_insert()/shiny::insertUI(), to lazily
  // load a module's UI on demand). Click handlers (.sidebar-label,
  // .settings-label, .tab-trigger) don't need this -- they're bound
  // delegated in sidebar.js/settings.js -- but moving each tab's settings
  // pane into place and selecting the first tab are one-off DOM
  // operations the page-ready setup only ever did for roots that existed
  // at that moment, so they need to be repeated explicitly for this one.
  Shiny.addCustomMessageHandler('bigdash-init-root', (msg) => {
      const id = msg.value;
      const $root = $(`.bigdash-app[data-bigdash-id="${id}"]`);
      if ($root.length === 0) return;

      moveSettingsForRoot(id, $root);
      initSettingsRoot(id);
      initSidebarRoot(id, $root);
  });
}
